arroyo-worker/src/connectors/filesystem/mod.rs

```diff
@@ -128,24 +130,79 @@ impl<K: Key, T: Data + Sync, R: MultiPartWriter<InputType = T> + Send + 'static>
             Arc::new(object_store),
             receiver,
             checkpoint_sender,
-            table,
+            table.clone(),
         );
         tokio::spawn(async move {
             writer.run().await.unwrap();
         });
+        let partition_func = get_partitioner_from_table(table);
         TwoPhaseCommitterOperator::new(Self {
             sender,
             checkpoint_receiver,
+            partitioner: partition_func,
             _ts: PhantomData,
         })
     }
 }
 
+fn get_partitioner_from_table<K: Key, T: Data + Serialize>(
+    table: FileSystemTable,
+) -> Option<Box<dyn Fn(&Record<K, T>) -> String + Send>> {
+    let Some(partitions) = table.file_settings.unwrap().partitioning else {
+        return None;
+    };
+    match (
+        partitions.time_partition_pattern,
+        partitions.partition_fields.is_empty(),
+    ) {
+        (None, false) => Some(Box::new(move |record: &Record<K, T>| {
+            partition_string_for_fields(&record.value, &partitions.partition_fields).unwrap()
+        })),
+        (None, true) => None,
+        (Some(TimePartitionPattern { pattern }), false) => {
+            Some(Box::new(move |record: &Record<K, T>| {
+                let time_partition = formatted_time_from_timestamp(record.timestamp, &pattern);
+                let field_partition =
+                    partition_string_for_fields(&record.value, &partitions.partition_fields)
+                        .unwrap();
+                format!("{}/{}", time_partition, field_partition)
+            }))
+        }
+        (Some(TimePartitionPattern { pattern }), true) => {
+            Some(Box::new(move |record: &Record<K, T>| {
+                formatted_time_from_timestamp(record.timestamp, &pattern)
+            }))
+        }
+    }
+}
+
+fn formatted_time_from_timestamp(timestamp: SystemTime, pattern: &str) -> String {
+    let datetime: DateTime<Utc> = DateTime::from(timestamp);
+    datetime.format(pattern).to_string()
+}
+fn partition_string_for_fields<T: Data + Serialize>(
+    value: &T,
+    partition_fields: &Vec<String>,
+) -> Result<String> {
+    let value = serde_json::to_value(value)?;
```

Comment:
this shouldn't fail (it would indicate a bug in codegen), so I would unwrap instead of propagating it